scalameta/semantic/shared/src/main/scala/scala/meta/semantic/Database.scala

```diff
@@ -0,0 +1,181 @@
+package scala.meta
+package semantic
+
+import java.io._
+import java.net.URI
+import java.util.zip._
+import scala.{Seq => _}
+import scala.collection.immutable.Seq
+import scala.collection.mutable
+import scala.compat.Platform.EOL
+import scala.util.Try
+import org.scalameta.data._
+import scala.meta.io._
+import scala.meta.internal.io.PlatformIO
+import scala.meta.internal.semantic.codecs._
+import scala.meta.internal.semantic.{proto => p}
+
+@data class Database(sources: Seq[AttributedSource]) {
+  locally {
+    val paths = sources.map(_.path.toString).toSet
+    if (sources.size != paths.size) {
+      val occs = mutable.Map[String, Int]().withDefault(_ => 0)
+      sources.foreach(source => occs(source.path.toString) += 1)
+      val duplicates = occs.filter(_._2 > 1).map(_._1)
+      sys.error("duplicate paths for " + occs.mkString(", "))
+    }
+  }
+
+  @deprecated("Use names instead", "1.8.0") def symbols: Map[Anchor, Symbol] = names
+  def names: Map[Anchor, Symbol] = sources.foldLeft(Map[Anchor, Symbol]())(_ ++ _.names)
+  def messages: Seq[Message] = sources.flatMap(_.messages)
+  def denotations: Map[Symbol, Denotation] = sources.foldLeft(Map[Symbol, Denotation]())(_ ++ _.denotations)
+
+  override def toString: String = sources.mkString(EOL + EOL)
+  def toBinary: Array[Byte] = this.toProto[p.Database].toByteArray
+  def writeToFile(dir: File): Unit = {
+    if (dir.exists) sys.error(s"implementation restriction: can't write a semantic database to an existing file")
+    val databaseRoot = Database.locateInClasspath(dir)
+    sources.foreach(source => source.writeToFile(new File(source.locateInDatabase(databaseRoot))))
+  }
+}
+
+object Database {
+  private[meta] def locateInClasspath(classpath: File): URI = {
+    if (classpath.isDirectory) {
+      new File(new File(classpath, "META-INF"), "semanticdb").toURI
+    } else if (classpath.getAbsolutePath.endsWith(".jar")) {
+      new URI("jar:" + classpath.toURI.toURL + "!META-INF/semanticdb/")
+    } else {
+      sys.error(s"unsupported classpath entry: $classpath")
+    }
+  }
+
+  def fromBinary(bytes: Array[Byte]): Database = p.Database.parseFrom(bytes).toMeta[Database]
```

Comment:
Later, I will write one megacomment about my thoughts wrt exceptions/monads for error handling.